src/audio/google_hotword_detect.c

```diff
@@ -0,0 +1,483 @@
+	ret = memcpy_s(comp, sizeof(*comp), comp_template,
+		       sizeof(struct sof_ipc_comp_process));
+	if (ret)
+		goto fail;
+
+	/* Create private component data */
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+		     sizeof(*cd));
+	if (!cd)
+		goto fail;
+	comp_set_drvdata(dev, cd);
+
+	/* Build component event */
+	ipc_build_comp_event(&cd->event, comp->comp.type, comp->comp.id);
+	cd->event.event_type = SOF_CTRL_EVENT_KD;
+	cd->event.num_elems = 0;
+
+	cd->msg = ipc_msg_init(cd->event.rhdr.hdr.cmd, sizeof(cd->event));
+	if (!cd->msg) {
+		comp_err(dev, "ghd_create(): ipc_msg_init failed");
+		goto fail;
+	}
+
+	/* Create component model data handler */
+	cd->model_handler = comp_data_blob_handler_new(dev);
+	if (!cd->model_handler) {
+		comp_err(dev, "ghd_create(): comp_data_blob_handler_new failed");
+		goto fail;
+	}
+
+	dev->state = COMP_STATE_READY;
+	comp_dbg(dev, "ghd_create(): Ready");
+	return dev;
+
+fail:
+	if (cd) {
+		comp_data_blob_handler_free(cd->model_handler);
+		ipc_msg_free(cd->msg);
+		rfree(cd);
+	}
+	if (dev)
+		rfree(dev);
+	return NULL;
+}
+
+static void ghd_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "ghd_free()");
+
+	comp_data_blob_handler_free(cd->model_handler);
+	ipc_msg_free(cd->msg);
+	rfree(cd);
+	rfree(dev);
+}
+
+static int ghd_params(struct comp_dev *dev,
+		      struct sof_ipc_stream_params *params)
+{
+	struct comp_buffer *sourceb;
+	int ret;
+
+	/* Detector is used only in KPB topology. It always requires channels
+	 * parameter set to 1.
+	 */
+	params->channels = 1;
+
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "ghd_params(): comp_verify_params failed.");
+		return -EINVAL;
+	}
+
+	/* This detector component will only ever have 1 source */
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+
+	if (sourceb->stream.channels != 1) {
+		comp_err(dev, "ghd_params(): Only single-channel supported");
+		return -EINVAL;
+	}
+
+	if (sourceb->stream.frame_fmt != SOF_IPC_FRAME_S16_LE) {
+		comp_err(dev, "ghd_params(): Only S16_LE supported");
+		return -EINVAL;
+	}
+
+	if (sourceb->stream.rate != KPB_SAMPLNG_FREQUENCY) {
+		comp_err(dev, "ghd_params(): Only 16KHz supported");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int ghd_setup_model(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	void *model;
+	size_t size = 0;
+	uint32_t crc = 0;
+	int ret;
+
+	model = comp_get_data_blob(cd->model_handler, &size, &crc);
+	if (!model || !size) {
+		comp_err(dev, "Model not set");
+		return -EINVAL;
+	}
+	comp_info(dev, "Model: data=0x%08x, size=%zu, crc=%u",
+		  (uint32_t)model, size, crc);
+
+	comp_info(dev, "GoogleHotwordVersion %d",
+		  GoogleHotwordVersion());
+
+	ret = GoogleHotwordDspInit(model);
+	cd->detected = 0;
+	cd->history_bytes = 0;
+	if (ret != 1) {
+		comp_err(dev, "GoogleHotwordDSPInit failed: %d", ret);
+		return -EINVAL;
```

Comment:
I'm assuming here hotword will return it's own error codes rather than errno.h versions ?

Reply:
correct